File: impls/rust.2/step4_if_fn_do.rs
extern crate rustyline;

use rustyline::error::ReadlineError;
use rustyline::Editor;

mod core;
mod env;
mod printer;
mod reader;
mod types;

use std::rc::Rc;
use types::MalValue::MalList;
use types::{MalError, MalResult, MalValue};

fn read(input: String) -> Result<Option<MalValue>, MalError> {
    return reader::Reader::read_str(input.to_string());
}

pub fn exec_fn(func: MalValue, args: Vec<MalValue>, env: Rc<env::Environment>) -> MalResult {
    match func {
        MalValue::MalFunction(f, _) => f(args),
        MalValue::MalFunc {
            eval,
            ref ast,
            ref env,
            ref params,
            ..
        } => {
            let a = &**ast;
            let fn_env = env::Environment::new(Some(env.clone()), params.as_vec(), Some(args));
            Ok(eval(a.clone(), Rc::new(fn_env))?)
        }
        _ => Err(MalError::EvalError(
            "Attempt to call non-function".to_string(),
        )),
    }
}

fn apply(ast: MalValue, env: Rc<env::Environment>) -> MalResult {
    match ast {
        MalValue::MalList(list) => {
            let evaluated_result = eval_ast(types::MalValue::MalList(list), env.clone())?;
            match evaluated_result {
                MalValue::MalList(list) => {
                    let func = list.first().unwrap();
                    let args = list.clone().split_off(1);

                    return exec_fn(func.clone(), args, env.clone());
                }
                _ => {
                    todo!("Handle eval_ast returning something that is not a list!");
                }
            }
        }
        _ => {
            return Err(MalError::EvalError(String::from(
                "apply called on non-list",
            )));
        }
    }
}

fn run_let(env: Rc<env::Environment>, bindings: MalValue, e: MalValue) -> MalResult {
    let new_env = Rc::new(env::Environment::new(Some(env.clone()), None, None));

    if let Some(binding_list) = bindings.as_vec() {
        assert!(binding_list.len() % 2 == 0);
        let mut key = binding_list.first().unwrap();

        for i in 0..binding_list.len() {
            if i % 2 == 0 {
                key = binding_list.get(i).unwrap();
            } else {
                new_env.set(
                    key.clone(),
                    eval(binding_list.get(i).unwrap().clone(), new_env.clone())?,
                )?;
            }
        }

        return eval(e, new_env);
    } else {
        return Err(MalError::EvalError(String::from("Missing bindings!")));
    }
}

fn eval(ast: MalValue, env: Rc<env::Environment>) -> MalResult {
    match ast.clone() {
        MalValue::MalList(list) => {
            if list.len() == 0 {
                return Ok(ast);
            }
            let first_element = &list[0];
            match first_element {
                MalValue::MalSymbol(ref s) if s == "def!" => {
                    return env.set(list[1].clone(), eval(list[2].clone(), env.clone())?);
                }
                MalValue::MalSymbol(ref s) if s == "fn*" => {
                    return Ok(MalValue::MalFunc {
                        eval: eval,
                        ast: Rc::new(list[2].clone()),
                        env: env,
                        params: Rc::new(list[1].clone()),
                    });
                }
                MalValue::MalSymbol(ref s) if s == "do" => {
                    for token in &list[1..list.len() - 1] {
                        eval(token.clone(), env.clone())?;
                    }

                    return eval(list.last().unwrap().clone(), env);
                }
                MalValue::MalSymbol(ref s) if s == "if" => {
                    assert!(list.len() >= 3);
                    let condition = eval(list[1].clone(), env.clone())?;
                    let true_expr = list[2].clone();
                    if condition.is_truthy() {
                        return eval(true_expr, env.clone());
                    } else if list.len() == 4 {
                        return eval(list[3].clone(), env.clone());
                    }
                    return Ok(MalValue::MalNil);
                }
                MalValue::MalSymbol(ref s) if s == "let*" => {
                    assert_eq!(3, list.len());
                    let bindings = list[1].clone();
                    let expr = list[2].clone();

                    let new_env = Rc::new(env::Environment::new(Some(env.clone()), None, None));

                    if let Some(binding_list) = bindings.as_vec() {
                        assert!(binding_list.len() % 2 == 0);
                        let mut key = binding_list.first().unwrap();
                        for i in 0..binding_list.len() {
                            if i % 2 == 0 {
                                key = binding_list.get(i).unwrap();
                            } else {
                                new_env.set(
                                    key.clone(),
                                    eval(binding_list.get(i).unwrap().clone(), new_env.clone())?,
                                )?;
                            }
                        }

                        return eval(expr, new_env);
                    } else {
                        return Err(MalError::EvalError(String::from("Missing bindings!")));
                    }
                }
                _ => match eval_ast(ast.clone(), env.clone())? {
                    MalList(list) => {
                        return exec_fn(list[0].clone(), list[1..].to_vec(), env);
                    }
                    _ => {
                        return Err(MalError::EvalError(String::from("Expected a list!")));
                    }
                },
            }
        }
        _ => {
            return eval_ast(ast, env);
        }
    }
}

fn eval_ast(ast: MalValue, env: Rc<env::Environment>) -> Result<MalValue, MalError> {
    match ast.clone() {
        MalValue::MalSymbol(s) => {
            if let Ok(func) = env.get(ast.clone()) {
                return Ok(func);
            }
            return Err(types::MalError::EvalError(format!(
                "Symbol '{}' not found",
                s
            )));
        }
        MalValue::MalList(list) => {
            let mut result = Vec::<types::MalValue>::new();

            for token in list {
                result.push(eval(token, env.clone())?);
            }

            return Ok(types::MalValue::MalList(result));
        }
        MalValue::MalVector(vector) => {
            let mut result = Vec::<types::MalValue>::new();

            for token in vector {
                result.push(eval(token, env.clone())?);
            }

            return Ok(types::MalValue::MalVector(result));
        }
        MalValue::MalHashmap(keys, values) => {
            let mut result = Vec::<types::MalValue>::new();

            for token in values {
                result.push(eval(token, env.clone())?);
            }

            return Ok(types::MalValue::MalHashmap(keys, result));
        }
        _ => {
            return Ok(ast);
        }
    }
}

fn print(input: MalValue) -> String {
    return crate::printer::pr_str(input, true);
}

fn rep(input: String, env: &Rc<env::Environment>) -> String {
    let ast = read(input);
    match ast {
        Err(e) => {
            return format!("Error! {:?}", e);
        }
        Ok(v) => match v {
            None => return String::default(),
            Some(a) => match eval(a, env.clone()) {
                Ok(result) => {
                    return print(result);
                }
                Err(e) => {
                    return format!("Error! {:?}", e);
                }
            },
        },
    }
}

fn main() {
    let mut rl = Editor::<()>::new();
    if rl.load_history(".mal-history").is_err() {
        println!("No previous history.");
    }

    let env = Rc::new(env::Environment::new(None, None, None));

    for (symbol, func) in crate::core::ns() {
        if let Err(e) = env.set(MalValue::MalSymbol(symbol.to_string()), func) {
            panic!("Failed add core item {}. {:#?}", symbol, e)
        }
    }

    rep("(def! not (fn* (a) (if a false true)))".to_string(), &env);

    loop {
        let readline = rl.readline("user> ");
        match readline {
            Ok(line) => {
                rl.add_history_entry(line.as_str());
                let result = rep(line.trim_end().to_string(), &env);
                print!("{}", result);
                println!();
            }
            Err(ReadlineError::Interrupted) => break,
            Err(ReadlineError::Eof) => break,
            Err(err) => {
                println!("Error: {:?}", err);
                break;
            }
        }
    }
}

File: impls/cc/step4_if_fn_do.cc
#include "core.hh"
#include "env.hh"
#include "printer.hh"
#include "reader.hh"
#include <iostream>
#include <map>
#include <string>

std::shared_ptr<MalType> eval(std::shared_ptr<MalType> input, Env *env);

std::shared_ptr<MalType> read(const std::string &input)
{
    return read_str(input);
}

std::shared_ptr<MalType> eval_ast(std::shared_ptr<MalType> ast, Env *env)
{
    switch (ast->type())
    {
    case MalType::Type::Symbol:
    {
        auto symbol = static_cast<MalSymbol &>(*ast);
        if (symbol.is_string() || symbol.is_keyword())
            return ast;

        return env->get(symbol);
    }
    case MalType::Type::List:
    {
        auto &list = static_cast<MalList &>(*ast);
        auto new_list = std::make_shared<MalList>(list.lparen(), list.rparen());
        for (auto &a : list)
        {
            auto l = eval(a, env);
            if (!l)
                return nullptr;
            new_list->push_back(l);
        }
        return new_list;
    }
    default:
        return ast;
    }
}

std::shared_ptr<MalType> eval(std::shared_ptr<MalType> input, Env *env)
{
    if (!input)
        return nullptr;

    if (input->type() != MalType::Type::List)
        return eval_ast(input, env);

    auto &list = static_cast<MalList &>(*input);
    if (list.empty())
        return input;

    if (!list.is_list())
        return eval_ast(input, env);

    if (list[0] && list[0]->type() == MalType::Type::Symbol)
    {
        std::string symbol = static_cast<const MalSymbol &>(*list[0]);
        if (symbol == "def!")
        {
            std::string key = static_cast<const MalSymbol &>(*list[1]);
            auto val = eval(list[2], env);
            if (!val)
                return nullptr;
            env->set(key, val);
            return val;
        }

        if (symbol == "let*")
        {
            Env new_env(env);
            auto &bindings = static_cast<const MalList &>(*list[1]);
            for (unsigned i = 0; i < bindings.size(); i += 2)
            {
                std::string key = static_cast<const MalSymbol &>(*bindings[i]);
                auto val = eval(bindings[i + 1], &new_env);
                if (!val)
                    return nullptr;
                new_env.set(key, val);
            }
            return eval(list[2], &new_env);
        }

        if (symbol == "do")
        {
            for (unsigned i = 1; i < list.size() - 1; ++i)
                eval(list[i], env);
            return eval(list[list.size() - 1], env);
        }

        if (symbol == "if")
        {
            auto cond = eval(list[1], env);
            if (cond && cond->type() == MalType::Type::Symbol)
            {
                auto condition = static_cast<MalSymbol &>(*cond);
                if (!condition)
                    return list.size() > 3 ? eval(list[3], env) : env->get("nil");
            }
            return eval(list[2], env);
        }

        if (symbol == "fn*")
        {
            auto closure = [list, env](std::vector<std::shared_ptr<MalType>> params)
            {
                // TODO: fix memory leak
                auto new_env = new Env(static_cast<const MalList &>(*list[1]), params, env);
                return eval(list[2], new_env);
            };
            return std::make_shared<MalFunc>(closure);
        }
    }

    auto plist = eval_ast(input, env);
    if (!plist)
        return nullptr;

    auto &new_list = static_cast<MalList &>(*plist);
    auto &func = static_cast<const MalFunc &>(*new_list[0]);

    std::vector<std::shared_ptr<MalType>> args;
    for (unsigned i = 1; i < new_list.size(); ++i)
        args.push_back(new_list[i]);

    return func(args);
}

std::string print(std::shared_ptr<MalType> input)
{
    return pr_str(input);
}

std::string rep(const std::string &input, Env *env)
{
    auto read_result = read(input);
    auto eval_result = eval(read_result, env);
    auto print_result = print(eval_result);
    return print_result;
}

int main(int argc, char *argv[])
{
    Env repl_env(nullptr);

    for (auto &[key, value] : ns())
        repl_env.set(key, value);

    rep("(def! not (fn* (a) (if a false true)))", &repl_env);

    while (!std::cin.eof())
    {
        std::string input;
        std::cout << "user> ";
        std::getline(std::cin, input);
        auto rep_result = rep(input, &repl_env);
        std::cout << rep_result << std::endl;
    }

    return 0;
}
